src/alfasim_sdk/result_reader/aggregator.py

```diff
@@ -1615,9 +1703,87 @@ def read_global_sensitivity_coefficients(
         return coefficients_dset[meta.qoi_index, meta.qoi_data_index]
+
+
+def read_history_matching_metadata(result_directory: Path) -> HistoryMatchingMetadata:
+    """
+    :param result_directory:
+        The directory to lookup for the History Matching result file.
+    """
+    return HistoryMatchingMetadata.from_result_directory(result_directory)
+
+
+def read_history_matching_result(
+    metadata: HistoryMatchingMetadata,
+    hm_type: Literal["deterministic", "probabilistic"],
+    hm_result_key: Optional[HistoryMatchingResultKeyType] = None,
+) -> Dict[HistoryMatchingResultKeyType, np.ndarray | float]:
+    """
+    :param metadata:
+        History Matching result metadata.
+    :param hm_type:
+        The type of HM analysis. Can be 'deterministic' or 'probabilistic'.
+    :param hm_result_key:
+        The id of the parametric vars to collect the result. Defaults to None, in which case the
+        result of all keys found in the metadata will be returned.
+    :return:
+        A dict mapping the HM result key (the parametric var id) to its corresponding result, which
+        could be an array with N values (N being the sampling size) for the probabilistic or a single
+        float for the deterministic.
+    """
+    if hm_type not in ("deterministic", "probabilistic"):
+        raise ValueError(f"history matching of type `{hm_type}` not supported")
+
+    if hm_type == "deterministic":
+        dataset_key = HISTORY_MATCHING_DETERMINISTIC_DSET_NAME
+    else:
+        assert hm_type == "probabilistic"
+        dataset_key = HISTORY_MATCHING_PROBABILISTIC_DSET_NAME
+
+    with open_result_file(metadata.result_directory) as result_file:
+        if not result_file:
+            return {}
+
+        result = result_file[HISTORY_MATCHING_GROUP_NAME][dataset_key]
+
+        result_map = {}
+        if hm_result_key is None:
+            for key, meta in metadata.hm_items.items():
+                result_map[key] = result[meta.data_index]
+        else:
+            meta = metadata.hm_items.get(hm_result_key)
+            if meta is not None:
+                result_map[hm_result_key] = result[meta.data_index]
+
+        return result_map
+
+
+@contextmanager
+def open_result_file(
+    result_directory: Path, result_filename: str = "result"
+) -> Iterator[Optional[h5py.File]]:
+    """
+    :param result_directory:
+        The directory to lookup for the result file.
+    :param result_filename:
+        The filename.
+    :return:
+        The result HDF file, or None if it doesn't exist or is still being created.
+    """
+    filepath = result_directory / result_filename
+    ignored_file = result_directory / f"{result_filename}.creating"
+
+    if not filepath.is_file():
+        yield None
+    # Avoid to read result files with incomplete metadata.
+    elif not ignored_file.is_file():
+        with _open_result_file(filepath) as file:
+            yield file
+    else:
+        yield None
```

Comment:
Minor, but I would keep each condition in their own `if`:

```suggestion
    filepath = result_directory / result_filename
    creating_file = result_directory / f"{result_filename}.creating"

    if not filepath.is_file():
        yield None
    # Do not open incomplete files.
    if creating_file.is_file():
        yield None
    else:
        with _open_result_file(filepath) as file:
            yield file
```

The way it is written is a bit more convoluted than it needs to be.